src/recordlinker/database/mpi_service.py

```diff
@@ -18,6 +18,64 @@
 from recordlinker import schemas
 
 
+def _filter_incorrect_blocks(
+        record: schemas.PIIRecord,
+        patients: typing.Sequence[models.Patient],
+        blocking_keys: list[str]
+) -> list[models.Patient]:
+    """
+    Filter a set of candidates returned via blocking from the MPI. The initial
+    SQL query returns a collection of candidates comprised of all patients in
+    the MPI belonging to a Person cluster for which at *least one* patient
+    satisfied blocking criteria. This function filters that candidate set to
+    include *only* those patients who either satisfied blocking criteria or
+    were missing a value for one or more blocked fields. This eliminates 
+    patients from consideration who have mismatched blocking information but
+    belonged to a Person cluster where a different record had correct blocking
+    values.
+
+    :param record: The PIIRecord of the incoming patient.
+    :param patients: The initial collection of candidates retrieved from the MPI.
+    :param blocking_keys: A list of strings of the fields used for blocking.
+    :returns: A filtered list of Patients from the MPI.
+    """
+    # Extract the acceptable blocking values from the incoming record
+    # Keys have already been getattr validated by caller, no need
+    # to check that they exist
+    blocking_vals_in_incoming = {}
+    for bk in blocking_keys:
+        key = getattr(models.BlockingKey, bk)
+        blocking_vals_in_incoming[bk] = [v for v in record.blocking_keys(key)]
```

Comment:
I think in the case where the valueset is empty, we want to exclude it from the dict.

Reply:
Yeah I think that's right. It starts to touch a little bit on the other half of missing information, which is missing fields in the incoming record (to be handled by our log odds wizardry), but since we'll be implementing that elsewhere this function won't need to be aware of missing incoming information; it can safely assume that if it's being run, the caller was satisfied that there's enough content to execute blocking.